witness/golang/omniwitness/usbarmory/storage/slots/journal.go

```diff
@@ -0,0 +1,258 @@
+func unmarshalEntry(r io.Reader) (*entry, error) {
+	e := &entry{}
+	if err := binary.Read(r, binary.BigEndian, &e.Magic); err != nil {
+		return nil, fmt.Errorf("failed to read magic: %v", err)
+	}
+	if string(e.Magic[:]) != magic0 {
+		return nil, fmt.Errorf("invalid header magic %v", e.Magic)
+	}
+	if err := binary.Read(r, binary.BigEndian, &e.Revision); err != nil {
+		return nil, fmt.Errorf("failed to read revision: %v", err)
+	}
+	if err := binary.Read(r, binary.BigEndian, &e.DataLen); err != nil {
+		return nil, fmt.Errorf("failed to read data length: %v", err)
+	}
+	if err := binary.Read(r, binary.BigEndian, &e.DataSHA256); err != nil {
+		return nil, fmt.Errorf("failed to read data SHA256: %v", err)
+	}
+	e.Data = make([]byte, e.DataLen)
+	if _, err := io.ReadFull(r, e.Data); err != nil {
+		return nil, fmt.Errorf("failed to read data; %v", err)
+	}
+	if h := sha256.Sum256(e.Data); !bytes.Equal(h[:], e.DataSHA256[:]) {
+		return e, fmt.Errorf("incorrect data SHA256 (%x), header claims (%x)", h, e.DataSHA256[:])
+	}
+	return e, nil
+}
+
+// marshalEntry serialises e and writes it to the provided writer.
+func marshalEntry(e *entry, w io.Writer) error {
+	if string(e.Magic[:]) != magic0 {
+		return fmt.Errorf("invalid header magic %v", e.Magic)
+	}
+	if h := sha256.Sum256(e.Data); !bytes.Equal(h[:], e.DataSHA256[:]) {
+		return fmt.Errorf("incorrect data SHA256 (%x), header claims (%x)", h, e.DataSHA256[:])
+	}
+
+	if err := binary.Write(w, binary.BigEndian, e.Magic); err != nil {
+		return fmt.Errorf("failed to write magic: %v", err)
+	}
+	if err := binary.Write(w, binary.BigEndian, e.Revision); err != nil {
+		return fmt.Errorf("failed to write revision: %v", err)
+	}
+	if err := binary.Write(w, binary.BigEndian, e.DataLen); err != nil {
+		return fmt.Errorf("failed to write data length: %v", err)
+	}
+	if err := binary.Write(w, binary.BigEndian, e.DataSHA256); err != nil {
+		return fmt.Errorf("failed to write data SHA256: %v", err)
+	}
+	if _, err := w.Write(e.Data); err != nil {
+		return fmt.Errorf("failed to write data; %v", err)
+	}
+	return nil
+}
+
+// blockReader provides an io.Reader wrapper for BlockReaderWriter instances.
+type blockReader struct {
+	dev BlockReaderWriter
+	buf []byte
+	pos uint
+}
+
+// newBlockReader creates a new reader for the given BlockReaderWriter, whose Read
+// function will start with the block address in lba.
+func newBlockReader(dev BlockReaderWriter, lba uint) *blockReader {
+	return &blockReader{
+		dev: dev,
+		buf: make([]byte, dev.BlockSize()),
+		pos: lba * dev.BlockSize(),
+	}
+}
+
+// Read implements io.Reader.
+func (br *blockReader) Read(b []byte) (int, error) {
+	if br.pos%br.dev.BlockSize() == 0 {
+		if err := br.dev.ReadBlocks(br.pos/br.dev.BlockSize(), br.buf); err != nil {
+			return 0, err
+		}
+	}
+	l := copy(b, br.buf[br.pos%br.dev.BlockSize():])
+	br.pos += uint(l)
+	return l, nil
+}
+
+func padToBlock(in []byte, blockSize uint) []byte {
```

Comment:
Is this necessary here? It looks like you already to the padding at a lower level?

Reply:
Good catch, I shifted that around a bit - removed this now.